llpc/context/llpcCompiler.cpp

```diff
@@ -1418,12 +1429,162 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+                                        cacheAccessor->getElfFromCache().codeSize);
+    } else {
+      // Cache miss.
+      if (partIdx == 0)
+        LLPC_OUTS("Fragment shader cache miss.\n");
+      else
+        LLPC_OUTS("Non fragment shader cache miss.\n");
+      // Compile the part pipeline.
+      // Keep the PipelineShaderInfo structs for the shaders we are including.
+      llvm::SmallVector<const PipelineShaderInfo *, 4> partPipelineShaderInfo;
+      for (const PipelineShaderInfo *oneShaderInfo : shaderInfo) {
+        if (oneShaderInfo && (partStageMask & (1 << oneShaderInfo->entryStage)))
+          partPipelineShaderInfo.push_back(oneShaderInfo);
+      }
+      result = buildPipelineInternal(context, partPipelineShaderInfo, PipelineLink::PartPipeline, &*elfLinkerPipeline,
+                                     &partPipelineBuffers[partIdx], stageCacheAccesses);
+      if (result != Result::Success)
+        break;
+      partPipelineElf = partPipelineBuffers[partIdx];
+
+      // If the "ELF" does not look like ELF, then it must be textual output from -emit-lgc, -emit-llvm, -filetype=asm.
+      // We can't link that, so just concatenate it on to the output.
+      if (partPipelineElf.size() < 4 || !partPipelineElf.startswith("\177ELF")) {
+        *pipelineElf += partPipelineElf;
+        continue;
+      }
+
+      // Store the part-pipeline ELF in the cache.
+      cacheAccessor->setElfInCache(BinaryData({partPipelineElf.size(), partPipelineElf.data()}));
+    }
+
+    if (Llpc::EnableOuts()) {                                                                                          \
+      ElfReader<Elf64> reader(m_gfxIp);
+      size_t readSize = 0;
+      if (reader.ReadFromBuffer(partPipelineElf.data(), &readSize) == Result::Success) {
+        LLPC_OUTS("===============================================================================\n");
+        LLPC_OUTS("// LLPC part-pipeline ELF (from cache or just compiled)\n");
+        LLPC_OUTS(reader);
+      }
+    }
+
+    if (!elfLinker) {
+      // Create the linker and its pipeline object.
+      elfLinkerPipeline.reset(context->getLgcContext()->createPipeline());
+      elfLinker.reset(elfLinkerPipeline->createElfLinker({}));
+
+      // Give selected pipeline state to the linker's pipeline object, required to complete the ELF link.
+      // We pass unlinked=true as we do not need to pass user data layout into the ELF link.
+      context->getPipelineContext()->setPipelineState(&*elfLinkerPipeline, /*hasher=*/nullptr, /*unlinked=*/true);
+    }
+
+    // Add the ELF to the linker.
+    elfLinker->addInputElf(llvm::MemoryBufferRef(partPipelineElf, ""));
+  }
+
+  if (result == Result::Success) {
+    // Link the two part-pipelines.
+    raw_svector_ostream elfStream(*pipelineElf);
+    bool ok = elfLinker->link(elfStream);
+    if (!ok) {
+      errs() << elfLinkerPipeline->getLastError() << "\n";
+      result = Result::ErrorUnavailable;
+      pipelineElf->clear();
```

Comment:
Should we be clearing at the start of the function?  What if the code fails earlier then the pipeline elf.  We should try to be consistent: if there is an error, then the pipeline elf will be an empty string, or if there is an error then the pipeline elf is undefined.  Either way, we should mention it in the function comments.

Reply:
Added a comment on top of the function.